### PR TITLE
feat(query/control): include variable memory limits for the controller

### PR DESCRIPTION
**Pull request by jsternberg**
The controller now supports setting an initial memory limit and setting
a maximum amount of memory that the controller may use separately from
the memory quota per query and the concurrency quota.

This allows the controller to increase the concurrency quota to a larger
number while setting the maximum amount of memory to a lower amount than
would be required for all queries to use 100% of their allowable memory.

Functionally, this means that a query will have a soft limit for an
initial memory byte quota that a query is guaranteed to have, a shared
pool that it is allowed access to in the case it uses more, and a hard
limit that no query may exceed to prevent runaway queries from taking
over the entire pool.

This change is completely backwards compatible with older configurations
as the new options will default to values that mimic the old behavior
where a query is allocated the full amount of its memory quota and the
maximum amount of memory is based on the concurrency quota and this
maximum memory quota.

In addition to the above, this also fixes a bug in the controller that
allowed it to run more than its concurrency as executing queries. This
happened when the results had finished being sent by the executor, but
the query had not yet been read and/or serialized. The executor would be
freed up and would take the next query even though the previous query
hadn't yet been finalized with `Done()`.

#15265

### CODE REVIEW

File: query/control/controller.go [hunk 1]
```diff
@@ -87,23 +100,53 @@ type Config struct {
 	ExecutorDependencies []flux.Dependency
 }
 
+// complete will fill in the defaults, validate the configuration, and
```
Comment (fchikwekwe) [inline query/control/controller.go]: Thanks for the descriptive commenting.

File: query/control/controller.go [hunk 2]
```diff
@@ -87,23 +100,53 @@ type Config struct {
...
 	if c.ConcurrencyQuota <= 0 {
 		return errors.New("ConcurrencyQuota must be positive")
 	}
 	if c.MemoryBytesQuotaPerQuery <= 0 {
 		return errors.New("MemoryBytesQuotaPerQuery must be positive")
 	}
+	if c.InitialMemoryBytesQuotaPerQuery <= 0 {
+		return errors.New("InitialMemoryBytesQuotaPerQuery must be positive")
+	}
+	if c.MaxMemoryBytes <= 0 {
```
Comment (fchikwekwe) [inline query/control/controller.go]: Looks like there may be an inconsistency between line 130 and line 139? They are checking the same value.

Comment (jsternberg) [reply query/control/controller.go]: Fixed. It seems like I accidentally made this same check twice.

File: query/control/memory.go [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+// RequestMemory will determine if the query can be given more memory
+// when it is requested.
+//
+// Note: This function accesses the memoryManager whose attributes
+// may be modified concurrently. Atomic operations are used to keep
+// it lockless. The data associated with this specific query are only
+// invoked from within a lock so they are safe to modify.
+// Second Note: The errors here are discarded anyway so don't worry
+// too much about the specific message or structure.
+func (q *queryMemoryManager) RequestMemory(want int64) (got int64, err error) {
```
Comment (nathanielc) [inline query/control/memory.go]: Does this method have a test case? I don't see where RequestMemory is used. Is that planned as a separate PR?

Comment (nathanielc) [reply query/control/memory.go]: I see its used on the Flux side of things https://github.com/influxdata/flux/pull/1922/files#diff-576d3093e5f774162ca06b1345b6ed07R191

**Pull request merged**